### PR TITLE
Bring windows segfault handler on par with linux 

### PR DESCRIPTION
**Pull request by Rocknest**
Adds a new way to dump stack traces on windows that closely resembles linux segfault handler. Works on i386, x86_64 (tested) and aarch64 (not tested).

I don't know where to put these definitions since there are no 'bits' for windows by architecture. 

Basically changes are only seen on i386. Test code: 
```zig
pub fn main() void {
    test1();
}

fn test1() void {
    illegal();
}

fn illegal() void {
    asm volatile ("nop");
    asm volatile ("ud2");
}
```

Before:
```
D:\zig>zig run ill.zig -target i386-windows-gnu
Illegal Instruction
```

After:
```diff
D:\zig>zig run ill.zig -target i386-windows-gnu
Illegal instruction at address 0xb6b034
D:\zig\ill.zig:10:5: 0xb6b034 in illegal (run.o)
    asm volatile ("nop");
    ^
D:\zig\ill.zig:6:12: 0xb64a17 in test1 (run.o)
    illegal();
           ^
D:\zig\ill.zig:2:10: 0xb52a97 in main (run.o)
    test1();
         ^
D:\zig\lib\zig\std\start.zig:125:65: 0xb519e2 in std.start.WinMainCRTStartup (run.o)                             
    std.os.windows.kernel32.ExitProcess(initEventLoopAndCallMain());
                                                                ^
???:?:?: 0x74f76358 in ??? (???)
???:?:?: 0x777d7b73 in ??? (???)
???:?:?: 0x777d7b43 in ??? (???)
!integer overflow
!D:\zig\lib\zig\std\debug.zig:140:65: 0xb6adbd in std.debug.dumpStackTraceFromBase (run.o)
!        printSourceAtAddress(debug_info, stderr, return_address - 1, tty_config) catch return;
!                                                                ^
!D:\zig\lib\zig\std\debug.zig:2331:31: 0xb649ca in std.debug.handleSegfaultWindowsExtra (run.o)
!        dumpStackTraceFromBase(regs.bp, regs.ip);
!                              ^
!D:\zig\lib\zig\std\debug.zig:2313:76: 0xb52a7a in std.debug.handleSegfaultWindows (run.o)
!        windows.EXCEPTION_ILLEGAL_INSTRUCTION => handleSegfaultWindowsExtra(info, 2, null),
!                                                                           ^
!???:?:?: 0x777dbba5 in ??? (???)
!???:?:?: 0x777d808a in ??? (???)
!???:?:?: 0x777e4215 in ??? (???)
!D:\zig\ill.zig:6:12: 0xb64a17 in test1 (run.o)
!    illegal();
!           ^
!D:\zig\ill.zig:2:10: 0xb52a97 in main (run.o)
!    test1();
!         ^
!D:\zig\lib\zig\std\start.zig:125:65: 0xb519e2 in std.start.WinMainCRTStartup (run.o)
!    std.os.windows.kernel32.ExitProcess(initEventLoopAndCallMain());
!                                                                ^
!???:?:?: 0x74f76358 in ??? (???)
!???:?:?: 0x777d7b73 in ??? (???)
!???:?:?: 0x777d7b43 in ??? (???)
```
Highlighted is a panic caused by underflow in `dumpStackTraceFromBase` when it received `0` from `StackIterator`. I'm sure its a bug.

There is also a problem that stack trace points to "nop" and not "ud2" instruction.

### CODE REVIEW

File: lib/std/debug.zig [hunk 1]
```diff
@@ -2307,16 +2307,284 @@ fn handleSegfaultLinux(sig: i32, info: *const os.siginfo_t, ctx_ptr: *const c_vo
...
-        windows.EXCEPTION_STACK_OVERFLOW => panicExtra(null, exception_address, "Stack Overflow", .{}),
+        windows.EXCEPTION_DATATYPE_MISALIGNMENT => handleSegfaultWindowsExtra(info, 0, "Unaligned Memory Access"),
+        windows.EXCEPTION_ACCESS_VIOLATION => handleSegfaultWindowsExtra(info, 1, null),
+        windows.EXCEPTION_ILLEGAL_INSTRUCTION => handleSegfaultWindowsExtra(info, 2, null),
+        windows.EXCEPTION_STACK_OVERFLOW => handleSegfaultWindowsExtra(info, 0, "Stack Overflow"),
         else => return windows.EXCEPTION_CONTINUE_SEARCH,
     }
 }
 
+// zig won't let me use an anon enum here
```
Comment (LemonBoy) [inline lib/std/debug.zig]: Why is that? What error does it give?

File: lib/std/debug.zig [hunk 2]
```diff
@@ -2307,16 +2307,284 @@ fn handleSegfaultLinux(sig: i32, info: *const os.siginfo_t, ctx_ptr: *const c_vo
...
+pub const windows_exception_context = switch (builtin.arch) {
+    .i386 => struct {
+        pub const haveContext = true;
+
+        pub fn getRegs(ptr: *c_void) struct {bp: usize, ip: usize} {
+            const ctx = @ptrCast(*const CONTEXT, @alignCast(@alignOf(CONTEXT), ptr));
+            return .{.bp = @intCast(usize, ctx.Ebp), .ip = @intCast(usize, ctx.Eip)};
+        }
+
+        pub const CONTEXT = extern struct {
```
Comment (LemonBoy) [inline lib/std/debug.zig]: All the structure definitions go in `os/windows/bits.zig`, the type aliases such as `WORD` and friends are already defined there. If you want to define those depending on the current architecture you can use some `usingnamespace` + `if`/`switch` like it's been done elsewhere in the stdlib.

File: lib/std/debug.zig (hunk 1 above)
Comment (Rocknest) [reply lib/std/debug.zig]:
```
D:\zig>zig run ill.zig -target x86_64-windows-gnu
.\lib\zig\std\debug.zig:2311:85: error: expected type 'std.debug.enum:2320:80', found 'std.debug.enum:2320:80'
        windows.EXCEPTION_DATATYPE_MISALIGNMENT => handleSegfaultWindowsExtra(info, .fmt, "Unaligned Memory Access"),
                                                                                    ^
.\lib\zig\std\debug.zig:2320:80: note: std.debug.enum:2320:80 declared here
fn handleSegfaultWindowsExtra(info: *windows.EXCEPTION_POINTERS, comptime msg: enum {fmt, segv, ill}, comptime format: ?[]const u8) noreturn {
```

Comment (LemonBoy) [reply lib/std/debug.zig]: Ah the infamous `enum` bug, I think it stops being a problem if you don't make it a comptime param.

**Comment by andrewrk** [comment]
> There is also a problem that stack trace points to "nop" and not "ud2" instruction.

Could this be related to 7336b750bd5e3c271a654e6d44d572cf0427a0b1?

File: lib/std/debug.zig (hunk 1 above)
Comment (Rocknest) [reply lib/std/debug.zig]: Nope, comptime does not make a difference

Comment (andrewrk) [reply lib/std/debug.zig]: Is there an issue for this? I'll take a look

Comment (LemonBoy) [reply lib/std/debug.zig]: #3707, the anonymous enum is analyzed twice and produces two disjoint types.

**Comment by Rocknest** [comment]
@andrewrk no, that function actually not even called once when i tested. 

I've managed to make it work, now it points to ud2 asm block. I have no idea why adding 1 to ip fixes it. Also stack iterator won't return nulls now.
```
D:\zig>zig run ill.zig -target i386-windows-gnu
Illegal instruction at address 0xdeb054
D:\zig\ill.zig:10:5: 0xdeb055 in illegal (run.o)
    asm volatile ("ud2");
    ^
D:\zig\ill.zig:6:12: 0xde4a17 in test1 (run.o)
    illegal();
           ^
D:\zig\ill.zig:2:10: 0xdd2a97 in main (run.o)
    test1();
         ^
D:\zig\lib\zig\std\start.zig:125:65: 0xdd19e2 in std.start.WinMainCRTStartup (run.o)
    std.os.windows.kernel32.ExitProcess(initEventLoopAndCallMain());
                                                                ^
???:?:?: 0x74f76358 in ??? (???)
???:?:?: 0x777d7b73 in ??? (???)
???:?:?: 0x777d7b43 in ??? (???)

D:\zig>
```

**Comment by andrewrk** [comment]
Thanks @Rocknest. This looks like a good change. Happy to merge after @LemonBoy's review comment https://github.com/ziglang/zig/pull/4319#discussion_r372501647 is solved

**Comment by LemonBoy** [comment]
> I've managed to make it work, now it points to ud2 asm block. I have no idea why adding 1 to ip fixes it. 

Hold your horses, this is not correct. Get a debugger and you'll see that the exception address is indeed correct and points to the `ud2` instruction, there's something fishy in the debug info and/or its handling.

**Comment by Rocknest** [comment]
@LemonBoy Yes the instruction is exactly where `ip` points to, but it looks like there is some extra -1 in printSourceAtAddress which is causing off by one error.

**Comment by LemonBoy** [comment]
> Yes the instruction is exactly where ip points to, but it looks like there is some extra -1 in printSourceAtAddress which is causing off by one error.

There's no -1 in `printSourceAtAddressWindows`, the `-1` offset is needed only when you're dealing with return addresses as you may be past the function boundary, but this is not the case as you have the ip.

**Comment by Rocknest** [comment]
@LemonBoy i suspect that off-by-one is debug.zig:472. `<=` instead `<`.

**Comment by LemonBoy** [comment]
>  i suspect that off-by-one is debug.zig:472. <= instead <.

If the comment above is to be trusted then yes, it should be `<`. I don't know too much about pdbs to say if that's correct or not, try it and see? :)

**Comment by andrewrk** [comment]
The comment is from a60ecdc6815555c59cd671e7846b1058be3f07fd

**Comment by Rocknest** [comment]
Looks like that's a proper fix. Stack traces look identical to linux (up to main fn).

**Comment by andrewrk** [comment]
:100:  Wonderful. Thank you both.

Nevermind the sr.ht build failure. It appears to be a temporary CI issue.

File: lib/std/os/windows/bits.zig [hunk 3]
```diff
@@ -887,9 +888,233 @@ pub const EXCEPTION_RECORD = extern struct {
...
+            Ebp: DWORD,
+            Eip: DWORD,
+            SegCs: DWORD,
+            EFlags: DWORD,
+            Esp: DWORD,
+            SegSs: DWORD,
+            ExtendedRegisters: [512]BYTE,
+
+            pub fn getRegs(ctx: *const CONTEXT) struct {bp: usize, ip: usize} {
+                return .{.bp = @intCast(usize, ctx.Ebp), .ip = @intCast(usize, ctx.Eip)};
```
Comment (andrewrk) [inline lib/std/os/windows/bits.zig]: `@intCast` is not needed here

File: lib/std/os/windows/bits.zig [hunk 4]
```diff
@@ -887,9 +888,233 @@ pub const EXCEPTION_RECORD = extern struct {
...
+            VectorRegister: [26]M128A,
+            VectorControl: DWORD64,
+            DebugControl: DWORD64,
+            LastBranchToRip: DWORD64,
+            LastBranchFromRip: DWORD64,
+            LastExceptionToRip: DWORD64,
+            LastExceptionFromRip: DWORD64,
+
+            pub fn getRegs(ctx: *const CONTEXT) struct {bp: usize, ip: usize} {
+                return .{.bp = @intCast(usize, ctx.Rbp), .ip = @intCast(usize, ctx.Rip)};
```
Comment (andrewrk) [inline lib/std/os/windows/bits.zig]: `@intCast` not needed here

File: lib/std/os/windows/bits.zig [hunk 5]
```diff
@@ -887,9 +888,233 @@ pub const EXCEPTION_RECORD = extern struct {
...
+            V: [32]NEON128,
+            Fpcr: DWORD,
+            Fpsr: DWORD,
+            Bcr: [8]DWORD,
+            Bvr: [8]DWORD64,
+            Wcr: [2]DWORD,
+            Wvr: [2]DWORD64,
+
+            pub fn getRegs(ctx: *const CONTEXT) struct {bp: usize, ip: usize} {
+                return .{.bp = @intCast(usize, ctx.DUMMYUNIONNAME.DUMMYSTRUCTNAME.Fp), .ip = @intCast(usize, ctx.Pc)};
```
Comment (andrewrk) [inline lib/std/os/windows/bits.zig]: `@intCast` not needed here

File: lib/std/os/windows/bits.zig (hunk 4 above)
Comment (Rocknest) [reply lib/std/os/windows/bits.zig]: I've got another nonsensical compile error
```
D:\zig>zig run ill.zig
.\lib\zig\std\os\windows\bits.zig:1036:29: error: array access of non-array type 'std.os.windows.bits.struct:1035:49'
                return .{ctx.Rbp, ctx.Rip};
                            ^
```

Comment (andrewrk) [reply lib/std/os/windows/bits.zig]: Generally the thing to do is to mark a workaround like this:

```zig
// TODO: The @as is a workaround for https://github.com/ziglang/zig/issues/4310
return .{.bp = @as(usize, ctx.Rbp), .ip = @as(usize, ctx.Rip)}; 
```

(replace 4310 with the issue number corresponding to this bug)

Does `@as` provide a workaround? That would be a preferable workaround to `@intCast`.

**Comment by andrewrk** [comment]
Nice work!

File: lib/std/os/windows/bits.zig (hunk 4 above)
Comment (Rocknest) [reply lib/std/os/windows/bits.zig]: Its a confusing compile error. Actually the problem is that anon list literal cannot coerce into given type.

I think it is https://github.com/ziglang/zig/issues/4148

**Pull request merged**